doc/associations.md

```diff
@@ -0,0 +1,233 @@
+  for the functionality of single association
+- [`"podio/detail/AssociationCollectionIterator.h"`](https://github.com/AIDASoft/podio/blob/master/include/podio/detail/AssociationCollectionIterator.h):
+  for the collection iterator functionality
+- [`"podio/detail/AssociationObj.h"`](https://github.com/AIDASoft/podio/blob/master/include/podio/detail/AssociationObj.h):
+  for the object layer functionality
+ - [`"podio/detail/AssociationCollectionData.h"`](https://github.com/AIDASoft/podio/blob/master/include/podio/detail/AssociationCollectionData.h):
+  for the collection data functionality
+- [`"podio/detail/AssociationFwd.h"`](https://github.com/AIDASoft/podio/blob/master/include/podio/detail/AssociationFwd.h):
+  for some type helper functionality and some forward declarations that are used
+  throughout the othe headers
+- [`"podio/detail/AssociationSIOBlock.h"`](https://github.com/AIDASoft/podio/blob/master/include/podio/detail/AssociationSIOBlock.h):
+  for defining the SIOBlocks that are necessary to use SIO
+  
+As is visible from this structure, we did not introduce an `AssociationData`
+class, since that would effectively just be a `float` wrapped inside a `struct`.
+  
+### Default and `Mutable` `Association` classes
+  
+A quick look into the `AssociationFwd.h` header will reveal that the default and
+`Mutable` `Association` classes are in fact just partial specialization of the
+`AssociationT` class that takes a `bool Mutable` as third template argument. The
+same approach is also followed by the `AssocationCollectionIterator`s:
+
+```cpp
+template<typename FromT, typename ToT, bool Mutable>
+class AssociationT;
+
+template <typename FromT, typename ToT>
+using Association = AssociationT<FromT, ToT, false>;
+
+template <typename FromT, typename ToT>
+using MutableAssociation = AssociationT<FromT, ToT, true>;
+```
+
+All podio generated datatypes have the following three public typedefs
+- `DefT` yields the default type
+- `MutT` yields the mutable type
+- `CollT` yields the collection type
+
+There are corresponding template helpers to retrieve these types in
+`AssocationFwd.h`. Note that these are not
+[*SFINAE*](https://en.cppreference.com/w/cpp/language/sfinae) friendly. However,
+since they are only used in contexts internally in non-SFINAE contexts, this
+doesn't really matter.
+
+Throught the implementation it is assumed that `FromT` and `ToT` always are the
+default types. This is ensured through `static_assert`s, resp. through usage of
+the aforementioned template helpers at the possible entry points. With this in
+mind, effectively all mutating operations on `Association`s are defined using
+SFINAE using the following template structure (taking here `setFrom` as an
+example)
+
+```cpp
+template <typename FromU,
+          typename = std::enable_if_t<Mutable && 
+                                      std::is_same_v<detail::GetDefT<FromU>, FromT>>>
+void setFrom(FromU value);
+```
```

Comment:
This needs a revision. Some of these have made it into mainline podio in a slightly different way since this was originally written